### DISCUSSION TITLE
PenaltyPeriodicSegmentalConstraint with different materials on the boundary

### DISCUSSION THREAD

**Discussion by sebastianpech** [idaholab/moose#28039] [Q&A General]
### Check these boxes if you have followed the posting rules.

- [X] Q&A General is the most appropriate section for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (screenshots as a last resort, triple back quotes around pasted text)

### Question

Hi,

I'm trying to do computational homogenization of a deformation problem by using the PenaltyPeriodicSegmentalConstraint. I use this instead of the periodic boundary condition constraint because my mesh is not necessarily periodic. I pretty much followed the test/tests/mortar/periodic_segmental_constraint/penalty_periodic_simple2d.i example from the documentation (https://mooseframework.inl.gov/source/constraints/PenaltyPeriodicSegmentalConstraint.html), but I'm running into the following problem.

In comparison to the example from above, I have material inclusions that touch the boundary. As I understand I can only supply one pair of primary_subdomain and secondary_subdomain for a PenaltyPeriodicSegmentalConstraint. However, for applying the material I need to distinguish the lower-level elements in the subdomain and I can't seem to find a way to do this.
I tried to use the boundary option of ComputeIsotropicElasticityTensor, however, this does not seem to work. Moose is still complaining about a missing material definition, so it seems that the newly generated subdomain is probably not in the sideset.

So my general problem (from my point of view) comes down to the following: I need all lower-level elements of one side of my specimen to be contained in one subdomain to use them in the PenaltyPeriodicSegmentalConstraint. However, In my material definition I need to be able to apply different materials to some of the elements in the subdomain. I can't seem to access them through a sideset definition and I can't split the subdomains because I need a single subdomain for the PenaltyPeriodicSegmentalConstraint. 
So is this even possible?

While reading through https://mooseframework.inl.gov/syntax/Materials/ I saw a note on "Evaluation of Material Properties on Element Faces". Maybe this could be an alternative approach, because the higher-level elements connected to the boundary have the material definition I need on the boundary. However, I couldn't find any further documentation on this.

**Comment by GiudGiud** [comment]
> I need all lower-level elements of one side of my specimen to be contained in one subdomain to use them in the PenaltyPeriodicSegmentalConstraint

>However, In my material definition I need to be able to apply different materials to some of the elements in the subdomain

This wont work with just block restriction of materials. 
You'll have to use a trick. For example a 3rd material that based on the current subdomain ids uses one or the other definition of a material

**Reply by sebastianpech** [reply]
Ok, thanks. So how would I approach this? Coding a custom material?

**Reply by GiudGiud** [reply]
most likely yes.
you use declareMaterialProperty to declare the new property
and you use getMaterialProperty in the constructor to retrieve references to the two material properties you are switching between
